common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -584,20 +596,28 @@ private static void notifyListener0(Future future, GenericFutureListener l) {
     }
 
     private void addListener0(GenericFutureListener<? extends Future<? super V>> listener) {
-        if (listeners == null) {
-            listeners = listener;
-        } else if (listeners instanceof DefaultFutureListeners) {
-            ((DefaultFutureListeners) listeners).add(listener);
+        if (this.listener == null) {
+            if (listeners == null) {
+                this.listener = listener;
+            } else {
+                listeners.add(listener);
+            }
         } else {
-            listeners = new DefaultFutureListeners((GenericFutureListener<?>) listeners, listener);
+            assert listeners == null;
+            listeners = new DefaultFutureListeners(this.listener, listener);
+            this.listener = null;
         }
     }
 
-    private void removeListener0(GenericFutureListener<? extends Future<? super V>> listener) {
-        if (listeners instanceof DefaultFutureListeners) {
-            ((DefaultFutureListeners) listeners).remove(listener);
-        } else if (listeners == listener) {
-            listeners = null;
+    private void removeListener0(GenericFutureListener<? extends Future<? super V>> toRemove) {
+        if (listener == toRemove) {
+            listener = null;
+        } else if (listeners != null) {
+            listeners.remove(toRemove);
+            // TODO we need ot compact it?
```

Comment:
@normanmaurer 
The compaction can happen while moving from 2 to 1 element too, but given that the listeners collection is already allocated won't worths

Reply:
Yeah, let's not bother with compaction. Removing listeners is so rare, and as you say, we already allocated the collection. Not to mention, if listeners are repeatedly added and removed, we don't want to thrash between the two modes and produce garbage on every switch.